### PR TITLE
Add machine type option memory-backend

### PR DESCRIPTION
**Pull request by huangyum**
Signed-off-by: Yumei Huang <yuhuang@redhat.com>
ID: 1926689

### CODE REVIEW

**Comment by huangyum** [comment]
Ran huge_page test loop with both qemu6.0 and qemu4.2, all cases passed with correct qemu cli.
@yfu Pls help review, thanks.

File: virttest/qemu_devices/qcontainer.py [hunk 1]
```diff
@@ -1438,6 +1438,15 @@ def machine_other(cmd=False):
                         cmd += ",accel=%s" % machine_accel
                     if machine_type_extra_params:
                         cmd += ",%s" % machine_type_extra_params.strip(',')
+                    machine_help = process.run("%s -M %s,\?" %
+                                               (self.__qemu_binary,
+                                                machine_type),
+                                               shell=True).stdout_text
+                    opt = "memory-backend"
+                    backend_id = "machine_backend"
```
Comment (yanan-fu) [inline virttest/qemu_devices/qcontainer.py:1446]: The id is fixed, how do you think about move it to base.cfg, make it clear for user to set params for the memory object ?

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -996,6 +996,22 @@ def add_memorys(devices, params):
                 for name in params.objects("mem_devs"):
                     dev = devices.memory_define_by_params(params, name)
                     devs.extend(dev)
+            machine_dev = devices.get_by_properties({"type": "machine"})[0]
+            machine_cmd = machine_dev.cmdline_nd()
+            output = re.findall(r",memory-backend=mem-([\w|-]+)", machine_cmd)
+            if output:
+                name = output[0]
+                params["size_%s" % name] = "%sM" % params["mem"]
+                params["use_mem_%s" % name] = "no"
```
Comment (yanan-fu) [inline virttest/qemu_vm.py:1005]: For this memory object for machine, can `use_mem` only be `no` ?

Comment (huangyum) [reply virttest/qemu_vm.py:1005]: Yes, it must be 'no', otherwise, a dimm device will generate.

File: virttest/qemu_devices/qcontainer.py (hunk 1 above)
Comment (huangyum) [reply virttest/qemu_devices/qcontainer.py]: Make sense. Added it to base.cfg, pls help review again.

File: virttest/qemu_vm.py (hunk 2 above)
Comment (yanan-fu) [reply virttest/qemu_vm.py:1005]: ok, thanks!

**Comment by yanan-fu** [comment]
@luckyh  Could you help review ? Thanks!
1. Check if use the new command in `machine_by_params`, add `memory-backend` option accordingly
2.  `add_memory` use new command or not depends on if `memory-backend` was added to machine.

For hugepage, some new change with the old version: remove the `extra_params += " -mem-path /mnt/kvm_hugepage"` in case cfg, add the string dev based on the hugepage_path in VT. (before, It is tricky, hardcode the mem-patch path in test case to use the same value with the hugepage_path in test_setup.HugePageConfig).

File: virttest/qemu_devices/qcontainer.py (hunk 1 above)
Comment (luckyh) [reply virttest/qemu_devices/qcontainer.py]: @yanan-fu what is the use case for user specified id?

File: virttest/qemu_devices/qcontainer.py [hunk 3]
```diff
@@ -1438,6 +1438,12 @@ def machine_other(cmd=False):
                         cmd += ",accel=%s" % machine_accel
                     if machine_type_extra_params:
                         cmd += ",%s" % machine_type_extra_params.strip(',')
+                    machine_help = self.execute_qemu("-M %s,?" %machine_type)
```
Comment (luckyh) [inline virttest/qemu_devices/qcontainer.py:1441]: @huangyum does it work well with non-x86 platforms?

File: virttest/qemu_devices/qcontainer.py (hunk 1 above)
Comment (yanan-fu) [reply virttest/qemu_devices/qcontainer.py]: The id is expected to bz fixed, and distinguish with other memory object. Do you have any other suggestion ?

File: virttest/qemu_devices/qcontainer.py (hunk 3 above)
Comment (huangyum) [reply virttest/qemu_devices/qcontainer.py:1441]: Yes, test passed on ppc host.

Comment (huangyum) [reply virttest/qemu_devices/qcontainer.py]: It works on arm and s390 platforms as well.

Comment (luckyh) [reply virttest/qemu_devices/qcontainer.py]: Get it, thanks!

File: virttest/qemu_vm.py [hunk 4]
```diff
@@ -996,6 +997,24 @@ def add_memorys(devices, params):
...
+                if params.get("hugepage_path"):
+                    backend_options["backend"] = "memory-backend-file"
+                    backend_options["mem-path"] = params["hugepage_path"]
+                backend_param = Params(backend_options)
+                dev = devices.memory_object_define_by_params(backend_param,
+                                                             name)
+                devs.append(dev)
+            elif params.get("hugepage_path"):
+                cmd = "-mem-path %s" % params["hugepage_path"]
+                devs.append(StrDev('extra', cmdline=cmd))
```
Comment (luckyh) [inline virttest/qemu_vm.py:1016]: Suggest the name `mem-path` instead of `extra`

Comment (huangyum) [reply virttest/qemu_vm.py]: Thanks, updated.  
Test passed on s390 host. 
@luckyh @yanan-fu Pls help review, thanks.

**Review by luckyh — APPROVED** [review]
LGTM

**Review by yanan-fu — APPROVED** [review]
LGTM

Acked-by: Yanan Fu <yfu@redhat.com>

**Comment by luckyh** [comment]
Thanks.

**Pull request merged**